packages/autoloader/src/class-version-selector.php

```diff
@@ -0,0 +1,92 @@
+<?php
+/* HEADER */ // phpcs:ignore
+
+/**
+ * Used to select package versions.
+ */
+class Version_Selector {
+
+	/**
+	 * Indicates whether a package class or file's plugin has a development version.
+	 *
+	 * @var Array
+	 */
+	private static $dev_plugin_in_use = array();
+
+	/**
+	 * Checks whether the selected package version should be updated.
+	 *
+	 * @param String $class The class name.
+	 * @param String $selected_version The currently selected package version.
+	 * @param String $compare_version The package version that is being evaluated to
+	 *                                determine if the version needs to be updated.
+	 * @param String $plugin The name of the plugin that contains the package.
+	 *
+	 * @return Boolean Returns true if the selected package version should be updated,
+	 *                 else false.
+	 */
+	public function update_package_version( $class, $selected_version, $compare_version, $plugin ) {
+
+		if ( is_null( $selected_version ) ) {
+			if ( $this->is_plugin_version_dev( $plugin ) ) {
+				self::$dev_plugin_in_use[ $class ] = true;
+			}
+
+			return true;
+		}
+
+		if ( $this->is_package_version_dev( $selected_version ) && self::$dev_plugin_in_use[ $class ] ) {
+			return false;
+		}
+
+		if ( $this->is_package_version_dev( $compare_version ) && $this->is_plugin_version_dev( $plugin ) ) {
+			self::$dev_plugin_in_use[ $class ] = true;
+			return true;
+		}
+
+		if ( version_compare( $selected_version, $compare_version, '<' ) ) {
```

Comment:
Judging by the logic, here you're expecting the `$compare_version` to never be a development version, right? If so, we can get a false positive if `99999-dev` gets through the previous conditions.

Reply:
Sorry - I didn't realize you were already looking at this PR, and I force pushed a few changes, including one that fixes this logic error. [This change](https://github.com/Automattic/jetpack/pull/15135/files#diff-ef9d1216cc594d83f995ac323d19b7b0R42) should properly handle situations where the package is a development version but the plugin is not.

Reply:
Hahah, OK, thanks! I couldn't help myself, this is very interesting stuff!